text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,121 @@
+  - Before this proposal, a Kafka administrator could set ACLs to Kafka topics so that a module can listen to only those tenants the module is enabled for in Okapi. When using topics with tenant collections, the ACLs can no longer be used on a tenant basis.
+
+## Rationale and Alternatives
+Here are other alternatives that were considered: 
+
+- **Merging All Topics Into A Single Topic**: This would be a maintenance nightmare. Messages could be consumed in error because so many consumer groups will be present and would rely on application code to filter out irrelevant messages for any particular module.
+- **Have A Single Topic For Each Module Area**: Module areas being Inventory, Acquisitions, Circulation etc. This shares the same pain as the previous point but to a lesser extent.
+- **Reduce Partition Count For Each Topic**: Partition counts should be tailored by a Kafka admin to suite the Kafka cluster involved; adjusting with load patterns and special cases that running applications might require. FOLIO has no guarantee that its configuration will be intact after its topics are initialized.<p> Reducing partition counts for each topic will produce decent impact on the problem illustrated by this document, but there is still a linear relationship between the number of tenants in a FOLIO cluster and the number of partitions in a Kafka cluster. Additional functionality, which require the use of more topics, will mean that partition counts in the Kafka cluster will grow with the addition of new tenants and release upgrades. Partition counts will be revisited sooner rather than later if only reducing the partition count of a topic rather than a topic's existence. The scalability of FOLIO is still bound by Kafka to lesser extent.<p> 
+Partitions of a topic can only be increased and not decreased, so the topic has to be dropped and recreated which should require some application disruption event.<p>
+This alternative is not included in this RFC because administration of a Kafka installation owned by a hosting provider is out of FOLIO's control. Additionally, setting alternate default partition counts in folio-kafka-wrapper does not require the rigor of an RFC.
+
+
+## Frequently Asked Questions
+- Should we ensure one tenant doesn't hog resources with a large number of messages which delays processing for other tenants with messages in a tenant collection topic?
+  - Within a tenant collection topic, there are still multiple partitions where messages will tend to be distributed evenly. So a tenant collection topic does not mean a strict queue with first in, first out. With that said, there are strategies that can be employed by a Kafka administrator when there is high offset lag or hot partitions in a cluster.
```

Comment:
The least favorable interpretation of the question could imply that a tenant collection topic has only one partition. Therefore when 10,000 events from tenant A and one event from tenant B is appended to the partition, processing has to occur for all 10,000 events before the one event from tenant B. Tenant A is "hogging" resources. That is what I mean by a tenant collection not being a strict queue. I have added some words pointing to one partition.

>Given tenant based topics and round robin processing, the consumer processes 1 event for tenant A, and then switches the topic and processes 1 event for tenant B.

I believe the single partition interpretation carries on this aspect of your example. The smallest unit that a consumer works with, is a partition. Round robin processing could be occurring on a partition basis not topics, I don't have the literature at my fingertips to determine what actually occurs. Two partitions from Tenant A could be processed before switching to tenant B. Another aspect could be: 100 messages in batch could be received for tenant A for processing before one message tenant B is processed.

>Given a tenant collection based topic, the consumer processes all 10000 events for tenant A, and then processes 1 event for tenant B. When we run 10 instances of the consumer and have 10 partitions each instance needs to handle 1000 tenant A events before the tenant B event can be processed.

1,000 events are processed before tenant B and not 10,000 events as prescribed by my least favorable interpretation. I think this is okay because:
- Asynchronous communication is involved. Near instant response is desired but not required. Measured slower response of the system during high load is to be expected. Other shared resources would be stressed.
- Burst of messages would not be a regular occurrence.
- Tenants with such regular high usage are usually moved into their own cluster.

But the fastest response will be achieved with tenant based topics in general. I have added a drawback to the relevant section. Please check it out!